Feat[WIP]: JIT compiled SQL expressions

jun0315 opened the pull request:
## Summary

MCS uses interepreted execution to calculate SQL expressions results. Here is an example of a SQL expression 'table1.col1 + FLOOR(table2.col1)'. Given that table1.col1 is DECIMAL and table2.col1 is DOUBLE there is a number of conditions that drives the calculation of this relatively simple example in runtime. Given that SQL types and expression tree are known before the query begins it is possible to replace interpretation with JIT to produce specialized compiled bytecode that is:

small
has no or almost no branches
optimized for the specific platform it is run at
This is a research mostly project which goal is to produce a set of microbenchmarks that:

leverages any JIT compiler available, e.g. LLVM, MIR
demonstrates a negative and positive effects of using JIT


## Related

[gsoc-2023](https://summerofcode.withgoogle.com/programs/2023/projects/hLufFI1y)

drrtuy committed:
Merge pull request #2766 from mariadb-corporation/build-dev6

Add ability to compile against MariaDB with new cost model

noorall committed:
[add] jit init

noorall committed:
[feat] compilation function of JIT module

noorall committed:
[feat] add implement of JITModuleMemoryManager

noorall committed:
[feat] add implement of deleteCompiledModule and registerExternalSymbol

noorall committed:
[feat] add implement of arena memory pool

jun0315 committed:
[feat] implement optimization functions for modules

[mod] add CMake config for llvm.

[feat] function

[feat] msc_jit

[fix] errors caused by namespace

[mod] resolved the build error

dr-m commented on jit/jit.cpp:
```diff
@@ -0,0 +1,467 @@
+  {
+    throw std::runtime_error("There is no compiled module with identifier");
+  }
+  module_identifier_to_memory_manager.erase(module_memory_manager_it);
+  compiled_code_size.fetch_sub(module.size, std::memory_order_relaxed);
```
Which code or data structure owns the memory that `module` referes to here? Could the memory actually be freed in the `erase` call? If so, then these statements should be swapped with each other to make the code safe.

dr-m commented on jit/jit.h:
```diff
@@ -0,0 +1,46 @@
+  std::unordered_map<uint64_t, std::unique_ptr<JITModuleMemoryManager>> module_identifier_to_memory_manager;
+  uint64_t current_module_key = 0;
+  std::atomic<size_t> compiled_code_size;
+  // TODO lock when compiling a module
+  mutable std::mutex jit_lock;
```
For any mutex or any potentially concurrently accessed data structure, it would be useful to document the protection rules. Many of these data members are actually protected by `jit_lock` in `deleteCompiledModule()`.

dr-m commented on jit/jit.h:
```diff
@@ -0,0 +1,46 @@
+  std::unique_ptr<JITSymbolResolver> symbol_resolver;
+  std::unordered_map<uint64_t, std::unique_ptr<JITModuleMemoryManager>> module_identifier_to_memory_manager;
+  uint64_t current_module_key = 0;
+  std::atomic<size_t> compiled_code_size;
```
I’d like to see comment for  documenting the lifetime and purpose of each data member. For `compiled_code_size` it would be useful to know what (if any) kind of padding or alignment overhead it would include.

Due to the lack of comments, it took me some time to find a counterpart of `deleteCompiledModule()` that would increment `compiled_code_size` and add something to `module_identifier_to_memory_manager`. It is `compileModule(std::unique_ptr<llvm::Module>)`. It seems to lack proper `jit_lock` protection.

dr-m commented on jit/jit.cpp:
```diff
@@ -0,0 +1,467 @@
+class Arena : private boost::noncopyable
+{
+ public:
+  Arena() : page_size(8)
+  {
+  }
```
We could live with an implicit default constructor if the initializer expression for `page_size` was revised to this value.

dr-m commented on jit/jit.cpp:
```diff
@@ -0,0 +1,467 @@
+    return allocated_size;
+  }
+
+  inline size_t getPageSize() const
+  {
+    return page_size;
+  }
+
+ private:
+  size_t page_size = 0;
```
I do not see any modification of this `page_size` in the code. I think that it could be declared simply as
```c++
  const size_t page_size = 8;
```
or even `static const` or `static constexpr` to save some memory, and then we could avoid explicitly declaring `Arena()`. I think that it is rather pointless to declare `const` data members `private` and to define public accessors for them.

What is the unit here? Bytes? Kilobytes? Some comments would be helpful.

Typically, `sysconf(PAGESIZE)` should return 4096 or a larger value.

dr-m commented on utils/funcexp/functor.h:
```diff
@@ -82,6 +84,14 @@ class Func
+  }
+  llvm::Value* compile()
+  {
+  }
```
It looks like the actual interesting part is still missing. The body of `compile()` seems to be missing `return nullptr;`.

dr-m commented on jit/compilefunction.cpp:
```diff
@@ -0,0 +1,23 @@
+{
+  llvm::IRBuilder<> b(module.getContext());
+  function->
+}
```
The member function body does not look like valid C++ to me.

dr-m commented on utils/funcexp/functor.h:
```diff
@@ -38,6 +38,8 @@
 
 #include "dataconvert.h"
 
+#include <llvm/IR/IRBuilder.h>
```
This is a rather large header file that will `#include` many other header files. I think that including that kind of a "jungle" in a rather popular header file (`functor.h` is being included from many other header files, which in turn could be included from many compilation units) could significantly include the compilation time.

We only seem to need this for the type name `llvm::Value`, which `compile()` would return a pointer to. I think that it would be better to define an opaque type and avoid including the header here. Something like this might work:
```c++
namespace llvm { class Value; }
```

dr-m reviewed (commented):
I understand that this is work in progress. Hopefully these comments are useful. It looks like we so far only have some interfaces but not any actual compilation, which would have been the main focus of my review.

I tried to compile this in the submodule `storage/columnstore/columnstore` under MariaDB/server@313c5a1dfb744aaef10586526dda89d2b4a50651. I got the build to complete, but the directory `storage/columnstore/columnstore/jit/CMakeFiles/jit.dir` in my `cmake -G Ninja` based build directory is empty. Other ColumnStore compilation units were being built.

jun0315 committed:
[feat] add JIT optimization interfaces for operator

[feat] adjust the structure of the compile method and implement IR for expression

[feat] add some testing code for JIT

[fix] arena memory allocation error issue

[mod] rollback the lib of llvm to the system built-in library

[feat] add JIT optimization for simpleColumnInt type

[feat] LogicOperator compile

jun0315 committed:
[feat] PredicateOperator compile

[mod] add isCompilable for column

[mod] add check for null row

[fix] the issue of compilation errors in the algorithm column IR

jun0315 committed:
[fix] jit add support for double and float type

[mod] remove some useless param

[mod] remove some useless file

[feat] add support for day() function

[feat] jit add support for func_month() function

dr-m commented on jit/compileoperator.cpp:
```diff
@@ -0,0 +1,92 @@
+  }
+  auto* func_type = llvm::FunctionType::get(return_type, {data_type, isNull_type}, false);
+  auto* func =
+      llvm::Function::Create(func_type, llvm::Function::ExternalLinkage, expression->alias(), module);
```
I would suggest
```c++
func->setDoesNotThrow();
```
so that no `.eh_frame` or `.rela.eh_frame` section will be generated.

noorall replied:
Done.

dr-m commented on jit/compileoperator.h:
```diff
@@ -0,0 +1,24 @@
+  JITCompiledOperator<int64_t> compiled_function_int64;
+  JITCompiledOperator<uint64_t> compiled_function_uint64;
+  JITCompiledOperator<double> compiled_function_double;
+  JITCompiledOperator<float> compiled_function_float;
```
Can multiple `compiled_function_` be set at a time?

noorall replied:
No, only one pointer will be set at a time.

dr-m commented on jit/expressionjit.h:
```diff
@@ -0,0 +1,105 @@
+  }
+
+ public:
+  int64_t getIntVal(rowgroup::Row& row, bool& isNull) override
+  {
+    return compiledOperator.compiled_function_int64(row.getData(), isNull);
+  }
+  uint64_t getUintVal(rowgroup::Row& row, bool& isNull) override
+  {
+    return compiledOperator.compiled_function_int64(row.getData(), isNull);
+  }
```
`compiled_function_uint64` seems to be unused. Do we really need this distinction? The ABI for signed and unsigned should be the same.

I would suggest to add some tests and to demonstrate that all this code is covered by the test cases.

dr-m commented on jit/expressionjit.h:
```diff
@@ -0,0 +1,105 @@
+  {
+    return compiledOperator.compiled_function_int64(row.getData(), isNull);
+  }
+  double getDoubleVal(rowgroup::Row& row, bool& isNull) override
+  {
+    return compiledOperator.compiled_function_double(row.getData(), isNull);
+  }
+  float getFloatVal(rowgroup::Row& row, bool& isNull) override
+  {
+    return compiledOperator.compiled_function_float(row.getData(), isNull);
+  }
```
Could we simply have one `compiled_function` member that will be `reinterpret_cast` to the appropriate type? That could avoid some run-time code duplication.

jun0315 committed:
[feat] func_day add support for call external function

[feat] func_day add support for timestamp

[feat] jit add external function timestampValueToInt

[mod] Modify the implementation of compileExpression

jun0315 committed:
[feat] jit supports more types of date conversions

[fix] the logic for func_hour

[feat] add the jit support for func_second

[mod] delete some useless define

jun0315 committed:
[fix] data type transform error

[mod] move jit to utils dir

[mod] remove link for NETSNMP_LIBRARIES

[mod] standardize the function name

drrtuy commented on CMakeLists.txt:
```diff
@@ -29,7 +29,7 @@ endif()
 INCLUDE(CheckCXXSourceCompiles)
 
-SET(CMAKE_CXX_STANDARD 11)
+SET(CMAKE_CXX_STANDARD 14)
```
We use c++20 if compiler handles it [via compiler flags](https://github.com/mariadb-corporation/mariadb-columnstore-engine/blob/develop/CMakeLists.txt#L53C1-L53C1). Is this set is necessary?

noorall replied:
It is necessary. If it is not specified as 14 during compilation, an error will be reported.

drrtuy commented on CMakeLists.txt:
```diff
@@ -358,6 +372,7 @@ ADD_SUBDIRECTORY(writeengine/splitter)
 ADD_SUBDIRECTORY(storage-manager)
 ADD_SUBDIRECTORY(datatypes)
 ADD_SUBDIRECTORY(tests)
+ADD_SUBDIRECTORY(jit)
```
This directory must be moved into ./utils and the paths changed accordingly.

noorall replied:
Done.

drrtuy commented on jit/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+
+add_library(jit SHARED ${jit_LIB_SRCS})
+
+target_link_libraries(jit ${NETSNMP_LIBRARIES} ${LLVM_LIBS} ${LLVM_SYSTEM_LIBS} ${LLVM_LDFLAGS})
```
You can drop ${NETSNMP_LIBRARIES} as this is an obsolete prerequisite.

noorall replied:
Done.

drrtuy commented on jit/compileoperator.cpp:
```diff
@@ -0,0 +1,105 @@
+{
+  auto columns = expression.get()->simpleColumnList();
+  llvm::IRBuilder<> b(module.getContext());
+  compileExternalFunction(module, b);
```
How many times compileExternalFunction will be actually called?

drrtuy commented on jit/compileoperator.cpp:
```diff
@@ -0,0 +1,105 @@
+  b.CreateRet(ret);
+
+  // Print the IR
+  llvm::verifyFunction(*func);
```
I suppose these two lines will go away.

noorall replied:
Done.

drrtuy commented on dbcon/execplan/arithmeticoperator.h:
```diff
@@ -210,6 +210,14 @@ class ArithmeticOperator : public Operator
+  inline llvm::Value* compile(llvm::IRBuilder<>& b, llvm::Value* data, llvm::Value* isNull,
+                              rowgroup::Row& row, ParseTree* lop, ParseTree* rop,
+                              CalpontSystemCatalog::ColDataType dataType) override;
+  inline llvm::Value* compileI(llvm::IRBuilder<>& b, llvm::Value* l, llvm::Value* r);
```
```suggestion
  inline llvm::Value* compileInt_(llvm::IRBuilder<>& b, llvm::Value* l, llvm::Value* r);
```

drrtuy commented on dbcon/execplan/arithmeticoperator.h:
```diff
@@ -210,6 +210,14 @@ class ArithmeticOperator : public Operator
+                              rowgroup::Row& row, ParseTree* lop, ParseTree* rop,
+                              CalpontSystemCatalog::ColDataType dataType) override;
+  inline llvm::Value* compileI(llvm::IRBuilder<>& b, llvm::Value* l, llvm::Value* r);
+  inline llvm::Value* compileF(llvm::IRBuilder<>& b, llvm::Value* l, llvm::Value* r);
```
```suggestion
  inline llvm::Value* compileFloat_(llvm::IRBuilder<>& b, llvm::Value* l, llvm::Value* r);
```

drrtuy commented on utils/rowgroup/rowgroup.cpp:
```diff
@@ -1043,6 +1043,103 @@ const CHARSET_INFO* Row::getCharset(uint32_t col) const
   return charsets[col];
 }
 
+void Row::compileIsNull(llvm::IRBuilder<>& b, llvm::Value* dataValue, llvm::Value* isNull, uint32_t colIndex)
```
I am curious why do you think this functionality belongs here and not to jit directory?

noorall replied:
The initial idea was to maintain consistency in function call behavior. It seems better to add a tool class in the jit directory to make the JIT code more centralized.

noorall replied:
> I am curious why do you think this functionality belongs here and not to jit directory?

I have moved it to a compilehelper class.

jun0315 committed:
[mod] move type compile function to compilehelper

add temp

add temp

[fix] some Werror

feat(cmake): hardcoded llvm include paths for a system llvm-13-dev @ubuntu20

WIP rebase on develop and change the code to compile with system llvm libs >= llvm-13

jun0315 committed:
llvm rpms for ci

no hardcoded llvm version

jun0315 committed:
llvm for deb

contos7 llvm cmake config

llvm conditional include

jun0315 committed:
__STDC_LIMIT_MACROS fix

add llvm deps

centos7 dep fix

jun0315 committed:
add libjit to deb packages

update llvm to 12 on Ubuntu 20 and to 16 on Debian11

WIP llvm-16 compilation issue

jun0315 committed:
WIP RowGroup helper class

WIP

WIP compilation fix

WIP introduce an expression name to disambigue JIT symbols in a JIT module

jun0315 committed:
WIP fixed cpimport rollback related issues in tests

WIP compilation fix

jun0315 committed:
WIP fix for SECOND() func

WIP avoid SECOND(TIMESTAMP) compilation

mariadb-LeonidFedorov commented:
work moved to https://github.com/mariadb-corporation/mariadb-columnstore-engine/pull/3165
